src/absil/ilread.fs

```diff
@@ -105,113 +106,79 @@ type BinaryFile() =
-                               int    _dwDesiredAccess, 
-                               int    _dwFileOffsetHigh,
-                               int    _dwFileOffsetLow, 
-                               SIZE_T _dwNumBytesToMap)
-
-    [<DllImport("kernel32", SetLastError=true)>]
-    extern bool UnmapViewOfFile (ADDR _lpBaseAddress)
-
-    let INVALID_HANDLE = new IntPtr(-1)
-    let MAP_READ    = 0x0004
-    let GENERIC_READ = 0x80000000
-    let NULL_HANDLE = IntPtr.Zero
-    let FILE_SHARE_NONE = 0x0000
-    let FILE_SHARE_READ = 0x0001
-    let FILE_SHARE_WRITE = 0x0002
-    let FILE_SHARE_READ_WRITE = 0x0003
-    let CREATE_ALWAYS  = 0x0002
-    let OPEN_EXISTING   = 0x0003
-    let OPEN_ALWAYS  = 0x0004
-
-type MemoryMappedFile(hMap: MemoryMapping.HANDLE, start:nativeint) =
+type MemMappedFile (fileName : string, memMappedFile : MemoryMappedFile) =
     inherit BinaryFile()
 
+    let accessor = memMappedFile.CreateViewAccessor (0L, 0L, MemoryMappedFileAccess.Read)
+
     static member Create fileName  =
-        //printf "fileName = %s\n" fileName
-        let hFile = MemoryMapping.CreateFile (fileName, MemoryMapping.GENERIC_READ, MemoryMapping.FILE_SHARE_READ_WRITE, IntPtr.Zero, MemoryMapping.OPEN_EXISTING, 0, IntPtr.Zero  )
-        //printf "hFile = %Lx\n" (hFile.ToInt64())
-        if ( hFile.Equals(MemoryMapping.INVALID_HANDLE) ) then
-            failwithf "CreateFile(0x%08x)" ( Marshal.GetHRForLastWin32Error() )
-        let protection = 0x00000002 (* ReadOnly *)
-        //printf "OK! hFile = %Lx\n" (hFile.ToInt64())
-        let hMap = MemoryMapping.CreateFileMapping (hFile, IntPtr.Zero, protection, 0,0, null )
-        ignore(MemoryMapping.CloseHandle(hFile))
-        if hMap.Equals(MemoryMapping.NULL_HANDLE) then
-            failwithf "CreateFileMapping(0x%08x)" ( Marshal.GetHRForLastWin32Error() )
-
-        let start = MemoryMapping.MapViewOfFile (hMap, MemoryMapping.MAP_READ,0,0,0n)
-
-        if start.Equals(IntPtr.Zero) then
-           failwithf "MapViewOfFile(0x%08x)" ( Marshal.GetHRForLastWin32Error() )
-        MemoryMappedFile(hMap, start)
-
-    member m.Addr (i:int) : nativeint = 
-        start + nativeint i
-
-    override m.ReadByte i = 
-        Marshal.ReadByte(m.Addr i)
-
-    override m.ReadBytes i len = 
+        let mmapFile =
+            // Use the FileStream-based API to open the file instead of using MemoryMappedFile.CreateFromFile()
+            // directly, so we can control the way the opened file is shared -- we want to allow other processes
+            // to be able to read the file while we're reading it, instead of opening it for exclusive access.
+            // Don't dispose the FileStream (e.g., with a 'use' binding); the MemoryMappedFile takes ownership
+            // of the stream once it's passed in and will take care of disposing it.
+            // TODO: Should we use FileShare.Read here instead of FileShare.ReadWrite (used to maintain the same functionality
```

Comment:
If you have a file open for read, another process can't open it for write. We don't need to open a file for read/write just to make sure another process doesn't modify the file underneath us.